KerbalStuff/celery.py

```diff
@@ -38,14 +36,45 @@ def send_mail(sender, recipients, subject, message, important=False):
             message['To'] = "undisclosed-recipients:;"
         else:
             message['To'] = ";".join(group)
-        print("Sending email from {} to {} recipients".format(sender, len(group)))
+        site_logger.info("Sending email from %s to %s recipients", sender, len(group))
         smtp.sendmail(sender, group, message.as_string())
     smtp.quit()
 
 
 @app.task
 def notify_ckan(mod_id, event_type):
-    if _cfg("notify-url") == "":
+    if not _cfg("notify-url"):
         return
+    import requests
     send_data = {'mod_id': mod_id, 'event_type': event_type}
     requests.post(_cfg("notify-url"), send_data)
+
+
+@app.task
+def update_from_github(working_directory, branch):
+    site_logger.info('Updating the site from github at: %s', working_directory)
+    try:
+        # pull new sources from git
+        from git import Repo, GitError
+        try:
+            repo = Repo(working_directory)
+            if repo.bare:
+                raise GitError()
+        except GitError:
+            site_logger.warning('No git repository at: %s', working_directory)
+            return
+        if repo.is_dirty():
```

Comment:
I think this will block us on alpha, the `systemd/spacedock.target` had to be modified to turn off all but one gunicorn service.

Reply:
If that's an issue then make a patch to disable the services on the systemd unit and apply it via the code.

Reply:
We may use `.gitattributes` to track different versions on unit files on different branches:
```
systemd/** merge=ours
```

Reply:
Or we could try `git pull --rebase` here, but depending on local changes this may not work and bring the working directory to un-merged state with the need to reset the branch.

Reply:
I don't think either of those ideas would make `repo.is_dirty()` return False when there are changes though, would they?

Reply:
Or would it in some cases? Let me check...

Reply:
Ah, `git stash` should do the trick!

Reply:
I can pull with changes when there aren't conflicts, but that may depend on some git setting I forgot about.

Reply:
Actually tried it -- no juice.
`git pull --rebase --autostash` failed, because there were _untracked_ files in worktree that were added in one of the remote commits. So it couldn't neither merge or rebase, and the changed files stayed in stash.

So if we want this to work robustly, we shouldn't allow pulling when the worktree is dirty. Meaning, we shouldn't make changes there and leave them be.

It's better if the service will autoupdate when all is well, and refuse to do so if something may complicate things. We can always make manual update in such a case.

Reply:
Maybe we should give up on `/etc/systemd/system/spacedock.target` being a symlink. Just copy and edit. Changes made to that file on the alpha branch wouldn't be reflected, but is that important?

Reply:
I agree; this particular unit should be pretty stable.

Reply:
OK, I'll make that change now.